CryptoAnalysis/src/main/java/crypto/ProviderDetection.java

```diff
@@ -0,0 +1,175 @@
+								String methodName = method.getName();
+									
+								SootClass methodRef = method.getDeclaringClass();
+								String refName = methodRef.toString();
+									
+								int parameterCount = method.getParameterCount();
+									
+								//list of JCA engine classes that are supported as CryptSL rules
+								String[] crySLRules = new String[] {"java.security.SecureRandom", "java.security.MessageDigest",
+																	"java.security.Signature", "javax.crypto.Cipher", 
+																	"javax.crypto.Mac", "javax.crypto.SecretKeyFactory",
+																	"javax.crypto.KeyGenerator", "java.security.KeyPairGenerator",
+																	"java.security.AlgorithmParameters", "java.security.KeyStore",
+																	"javax.net.ssl.KeyManagerFactory", "javax.net.ssl.SSLContext",
+																	"javax.net.ssl.TrustManagerFactory"};
+									
+								boolean ruleFound = Arrays.asList(crySLRules).contains(refName);
+									
+								if((ruleFound) && (methodName.matches("getInstance")) && (parameterCount==2) ) {
+									Value vl = exp.getArg(1);
+									String strType = getProviderType(vl);
+										
+									if(strType.matches("java.security.Provider")) {
+										String provider = getProvider(stmt, method, vl, icfg);
+										rules = chooseRules(rules, provider, refName); 
+									}
+										
+									else if (strType.matches("java.lang.String")) {
+										String provider = vl.toString();
+										rules = chooseRules(rules, provider, refName);
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+		
+		return rules;
+	}
+	
+	//methods used from the "doAnalysis() method"
+	//-----------------------------------------------------------------------------------------------------------------
+	
+	private String getProviderType(Value vl) {
+		Type type = vl.getType();
+		String strType = type.toString();
+		return strType;
+	}
+	
+	
+	private String getProvider(JAssignStmt stmt, SootMethod method, Value vl, JimpleBasedInterproceduralCFG icfg) {
+		String provider = null;
+		
+		BackwardQuery query = new BackwardQuery(new Statement(stmt,method), new Val(vl, method));
+		
+		//Create a Boomerang solver.
+		Boomerang solver = new Boomerang(new DefaultBoomerangOptions(){
+			public boolean onTheFlyCallGraph() {
+				//Must be turned of if no SeedFactory is specified.
+				return false;
+			};
+		}) {
+			@Override
+			public BiDiInterproceduralCFG<Unit, SootMethod> icfg() {
+				return icfg;
+			}
+
+			@Override
+			public boomerang.seedfactory.SeedFactory<NoWeight> getSeedFactory() {
+				return null;
+			}
+		};
+		
+		//Submit query to the solver.
+		BackwardBoomerangResults<NoWeight> backwardQueryResults = solver.solve(query);
+		solver.debugOutput();
+		
+		Map<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> map = backwardQueryResults.getAllocationSites();
+		Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry = map.entrySet().iterator().next();
```

Comment:
That's a complicated way of iterating over a map.

You can use 
`for(entry : map.entrySet()){}`

Reply:
Also you just are just using the first entry. What about the others?

Reply:
I tested this scenario using Boomerang and the size of the _map_ is always 1 since there is only one allocation site for the _Provider_ variable. I also just tested your suggestion about the iteration and the output is the same since there is always only 1 element in the _map_.  In my case, I was just getting the first entry from the map, and extracting the corresponding _key_ value. Anyway, I can use your iteration if you want. Thanks

Reply:
A code such as 
`Cipher.getInstance(...,(someChoice() ? "BouncyCastle" : "Default"))`
would yield a size of the map larger than one. And your code *randomly* picks one of the providers as the order of the elements in the map depend on the hashcode of the objects, which may differ in each run of the JVM. We have to prevent such non-deterministic choices.

Please think of a solution and better throw an exception in the case there are multiple providers. Then we at least know where the code crashes. Thanks.

Reply:
@johspaeth I had not thought about this case. Thanks for exposing this scenario. I will make the required changes to the code.